defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+package defrag
+
+import (
+	"container/list"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"code.google.com/p/gopacket"
+	"code.google.com/p/gopacket/layers"
+)
+
+// Quick and Easy yo use debug code to trace
+// how defrag works.
+var debug debugging = false // or flip to false
+type debugging bool
+
+func (d debugging) Printf(format string, args ...interface{}) {
+	if d {
+		log.Printf(format, args...)
+	}
+}
+
+type Defragmenter interface {
```

Comment:
Re: wishlist:  sorry about that.  My attempt with that was to show what functions/args the type should provide, not to actually use an interface.

I think it would be best to work like https://godoc.org/github.com/google/gopacket/tcpassembly#Assembler does.  Don't worry about a high-level interface here... just create a type with a small number of exported functions/vars.  If folks want, they can create high-level interfaces for that type on their own.

I think that with Go, a high-level interface per package actually isn't necessary or particularly useful.  In any language with a hierarchical inheritance model, it's hugely necessary, since otherwise users can't implement their own implementations.  But in Go, because of duck-typing, if I want to make my code modular I can write the interface to match the implementation.  So if I want to, for example, inject test code, I can write my own interface with just the functions I care about, pass in the production class in normal code, then pass in my own test code in a test.

A good example of this approach is gopacket.PacketSource, pcap.Handle, afpacket.TPacket, etc.  The packages that provide implementations (pcap.Handle, afpacket.TPacket) don't expose interfaces... they just give you types you can use.  But the system that needs an interface (gopacket) creates one (PacketSource) where/when it needs to.